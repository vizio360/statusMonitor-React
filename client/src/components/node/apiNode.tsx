import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {Node, NodeStatus, INodeEvents} from './';
import {EndpointOptions, Defaults, jsPlumb, jsPlumbInstance} from 'jsPlumb';
import apiImage from '@images/api.png';
import {Fetch, IStatus, IService} from '@app/fetchData';

interface IAPINode {
  jsPlumb: jsPlumbInstance;
  service: IService;
  events: INodeEvents;
}

class APINode extends React.Component<IAPINode> {
  render() {
    return (
      <Node
        id={this.props.service.id}
        name={this.props.service.name}
        jsPlumb={this.props.jsPlumb}
        image={apiImage}
        status={NodeStatus.HEALTHY}
        events={this.props.events}
        uiProps={this.props.service.uiProps}
      />
    );
  }
}

export default APINode;
