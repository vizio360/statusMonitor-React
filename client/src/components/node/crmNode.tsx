import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {INodeEvents, Node, NodeStatus} from './';
import {EndpointOptions, Defaults, jsPlumb, jsPlumbInstance} from 'jsPlumb';
import crmImage from '@images/crm.png';
import {Fetch, IStatus, IService} from '@app/fetchData';

interface ICRMNode {
  jsPlumb: jsPlumbInstance;
  service: IService;
  events?: INodeEvents;
}

class CRMNode extends React.Component<ICRMNode> {
  constructor(props: ICRMNode) {
    super(props);
  }
  componentDidMount() {
    console.log('mounted!');
  }

  componentWillUnmount() {
    console.log('umounted');
  }

  componentWillReceiveProps(props: ICRMNode) {
    console.log(props);
  }

  render() {
    return (
      <Fetch service={this.props.service}>
        {(status: IStatus) => (
          <Node
            id={this.props.service.id}
            name={this.props.service.name}
            jsPlumb={this.props.jsPlumb}
            image={crmImage}
            status={
              status.status === 'Unhealthy'
                ? NodeStatus.UNHEALTHY
                : NodeStatus.HEALTHY
            }
            events={this.props.events}
            uiProps={this.props.service.uiProps}
          />
        )}
      </Fetch>
    );
  }
}

export default CRMNode;
