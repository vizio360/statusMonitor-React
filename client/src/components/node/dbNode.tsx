import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {Node, NodeStatus, INodeEvents} from './';
import {EndpointOptions, Defaults, jsPlumb, jsPlumbInstance} from 'jsPlumb';
import dbImage from '@images/db.png';
import {Fetch, IStatus, IService} from '@app/fetchData';

interface IDBNode {
  jsPlumb: jsPlumbInstance;
  service: IService;
  events: INodeEvents;
}

class DBNode extends React.Component<IDBNode> {
  componentDidMount() {}

  render() {
    return (
      <Node
        id={this.props.service.id}
        name={this.props.service.name}
        jsPlumb={this.props.jsPlumb}
        image={dbImage}
        status={NodeStatus.UNHEALTHY}
        events={this.props.events}
        uiProps={this.props.service.uiProps}
      />
    );
  }
}

export default DBNode;
