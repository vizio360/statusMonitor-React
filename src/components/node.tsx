import * as React from 'react';
import * as ReactDOM from 'react-dom';

interface INodeProps {
  image: Element;
  name: string;
}

class Node extends React.Component<INodeProps> {}

export default Node;
