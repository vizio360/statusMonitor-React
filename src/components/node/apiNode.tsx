import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {Node, NodeStatus, INodeEvents} from './';
import {EndpointOptions, Defaults, jsPlumb, jsPlumbInstance} from 'jsPlumb';
import apiImage from '../../../images/api.png';
import {Fetch, IStatus, IService} from '../../fetchData';

interface IAPINode {
  jsPlumb: jsPlumbInstance;
  service: IService;
  events: INodeEvents;
}

class APINode extends React.Component<IAPINode> {
  render() {
    return (
      <Fetch service={this.props.service}>
        {(status: IStatus) => (
          <Node
            id={this.props.service.id}
            name={this.props.service.name}
            jsPlumb={this.props.jsPlumb}
            image={apiImage}
            status={
              status.status === 'Unhealthy'
                ? NodeStatus.UNHEALTHY
                : NodeStatus.HEALTHY
            }
            events={this.props.events}
            uiProps={this.props.service.uiProps}
          />
        )}
      </Fetch>
    );
  }
}

export default APINode;
