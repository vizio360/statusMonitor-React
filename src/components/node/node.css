.node {
  width: 100px;
  height: 100px;
  position: absolute;
  text-align: center;
}
.nodeOverlay {
  top: 0;
  left: 0;
  width: 100px;
  height: 100px;
  position: absolute;
}
