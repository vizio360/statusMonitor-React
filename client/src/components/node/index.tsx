import * as React from 'react';
import * as ReactDOM from 'react-dom';
import './node.css';
import fireImage from '@images/fire.gif';
import {
  Status,
  ServiceType,
  IService,
  IServiceLastKnownState,
} from '@dataTypes';

import crmImage from '@images/crm.png';
import dbImage from '@images/db.png';
import apiImage from '@images/api.png';

interface INodeEvents {
  onDoubleClick: (id: string) => void;
}

interface INodeProps {
  service: IService;
  serviceState: IServiceLastKnownState;
  className?: string;
  events?: INodeEvents;
}

class Node extends React.Component<INodeProps> {
  nodeEvents: INodeEvents;

  constructor(props: INodeProps) {
    super(props);
    this.nodeEvents = props.events;
    this.onDoubleClick = this.onDoubleClick.bind(this);
  }

  onDoubleClick(e: React.MouseEvent<HTMLElement>) {
    if (this.nodeEvents && this.nodeEvents.onDoubleClick)
      this.nodeEvents.onDoubleClick(this.props.service.id);
  }

  getNodeTypeImage(serviceType: ServiceType) {
    let img: string;
    switch (serviceType) {
      case ServiceType.DB:
        img = dbImage;
        break;
      case ServiceType.API:
        img = apiImage;
        break;
      case ServiceType.CRM:
        img = crmImage;
        break;
    }
    return img;
  }

  render() {
    return (
      <div
        id={this.props.service.id}
        className={this.props.className ? this.props.className : 'node'}
        style={this.props.service.uiProps}
        onDoubleClick={this.onDoubleClick}>
        <img
          src={this.getNodeTypeImage(this.props.service.type)}
          width="100%"
          height="100%"
        />
        {this.props.serviceState.status !== Status.HEALTHY ? (
          <img
            className="nodeOverlay"
            src={fireImage}
            width="100%"
            height="100%"
          />
        ) : (
          ''
        )}
        <p>{this.props.service.name}</p>
      </div>
    );
  }
}

export {Node, INodeEvents};
