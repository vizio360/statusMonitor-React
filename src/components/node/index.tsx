import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {
  DragEventCallbackOptions,
  EndpointOptions,
  Defaults,
  jsPlumb,
  jsPlumbInstance,
} from 'jsPlumb';
import './node.css';
import fireImage from '../../../images/fire.gif';

interface INodeEvents {
  stopDrag: (params: DragEventCallbackOptions) => void;
  onDoubleClick: (id: string) => void;
}

interface INodeProps {
  jsPlumb: jsPlumbInstance;
  image: any;
  name: string;
  id: string;
  status: NodeStatus;
  className?: string;
  events?: INodeEvents;
  uiProps: {top: number; left: number};
}

enum NodeStatus {
  HEALTHY,
  UNHEALTHY,
}

enum NodeType {
  CRM = 'CRM',
  API = 'API',
  DB = 'DB',
}

class Node extends React.Component<INodeProps> {
  jsPlumb: jsPlumbInstance;
  nodeEvents: INodeEvents;

  constructor(props: INodeProps) {
    super(props);
    this.jsPlumb = props.jsPlumb;
    this.nodeEvents = props.events;
    this.onDragStop = this.onDragStop.bind(this);
    this.onDoubleClick = this.onDoubleClick.bind(this);
  }

  getTargetEndPoint(id: string) {
    let inEndPointOptions: EndpointOptions = {
      maxConnections: 10,
      anchor: 'Top',
      paintStyle: {fill: '#F00'},
      type: 'Dot',
      id: id + 'target',
      scope: 'dotEndPoint',
      reattachConnections: true,
      parameters: {},
      isTarget: true,
    };
    return inEndPointOptions;
  }

  getSourceEndPoint(id: string) {
    let outEndPointOptions: EndpointOptions = {
      maxConnections: 10,
      anchor: 'Bottom',
      paintStyle: {fill: '#00F'},
      type: 'Dot',
      id: id + 'source',
      scope: 'dotEndPoint',
      reattachConnections: true,
      parameters: {},
      isSource: true,
    };
    return outEndPointOptions;
  }

  onDragStop(params: DragEventCallbackOptions) {
    if (this.nodeEvents && this.nodeEvents.stopDrag)
      this.nodeEvents.stopDrag(params);
  }

  onDoubleClick(e: React.MouseEvent<HTMLElement>) {
    if (this.nodeEvents && this.nodeEvents.onDoubleClick)
      this.nodeEvents.onDoubleClick(this.props.id);
  }

  componentDidMount() {
    this.jsPlumb.addEndpoint(
      this.props.id,
      this.getSourceEndPoint(this.props.id),
    );
    this.jsPlumb.addEndpoint(
      this.props.id,
      this.getTargetEndPoint(this.props.id),
    );
    this.jsPlumb.draggable(this.props.id, {stop: this.onDragStop});
  }

  render() {
    return (
      <div
        id={this.props.id}
        className={this.props.className ? this.props.className : 'node'}
        style={this.props.uiProps}
        onDoubleClick={this.onDoubleClick}>
        <img src={this.props.image} width="100%" height="100%" />
        {this.props.status !== NodeStatus.HEALTHY ? (
          <img
            className="nodeOverlay"
            src={fireImage}
            width="100%"
            height="100%"
          />
        ) : (
          ''
        )}
        <p>{this.props.name}</p>
      </div>
    );
  }
}

export {Node, NodeStatus, NodeType, INodeEvents};
