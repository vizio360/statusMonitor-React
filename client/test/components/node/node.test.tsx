import * as React from 'react';
import * as ReactDOM from 'react-dom';
import {Node} from '@app/components/node';
import renderer from 'react-test-renderer';
import {
  Status,
  IService,
  IServiceLastKnownState,
  ServiceType,
} from '@dataTypes';

describe('Node', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const service: IService = {
    id: '1',
    name: 'A_SERVICE',
    uri: 'http://service',
    timeout: 30,
    type: ServiceType.DB,
    categories: [],
    uiProps: {
      top: 10,
      left: 10,
    },
  };

  const serviceState: IServiceLastKnownState = {
    serviceId: '1',
    status: Status.HEALTHY,
    responseBody: 'some json body',
  };

  const jsx = (
    <Node key={service.id} service={service} serviceState={serviceState} />
  );

  it('renders a node', () => {
    let node = renderer.create(jsx);
    let tree = node.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('re-renders a node if service status changes', () => {
    let node = renderer.create(jsx);
    let tree = node.toJSON();
    expect(tree).toMatchSnapshot();

    const serviceStateUpdated: IServiceLastKnownState = {
      serviceId: '1',
      status: Status.UNHEALTHY,
      responseBody: 'some json body',
    };
    let jsxUpdate = (
      <Node
        key={service.id}
        service={service}
        serviceState={serviceStateUpdated}
      />
    );
    node.update(jsxUpdate);
    tree = node.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('re-renders a node if service ui properties change', () => {
    let node = renderer.create(jsx);
    let tree = node.toJSON();
    expect(tree).toMatchSnapshot();

    let updatedService: IService = JSON.parse(JSON.stringify(service));
    updatedService.uiProps = {
      top: 500,
      left: 600,
    };
    let jsxUpdate = (
      <Node
        key={service.id}
        service={updatedService}
        serviceState={serviceState}
      />
    );
    node.update(jsxUpdate);
    tree = node.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
