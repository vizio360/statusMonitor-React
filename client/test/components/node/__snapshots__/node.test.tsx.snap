// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node re-renders a node if service status changes 1`] = `
<div
  className="node"
  id="1"
  onDoubleClick={[Function]}
  style={
    Object {
      "left": 10,
      "top": 10,
    }
  }
>
  <img
    height="100%"
    src="test-file-stub"
    width="100%"
  />
  
  <p>
    A_SERVICE
  </p>
</div>
`;

exports[`Node re-renders a node if service status changes 2`] = `
<div
  className="node"
  id="1"
  onDoubleClick={[Function]}
  style={
    Object {
      "left": 10,
      "top": 10,
    }
  }
>
  <img
    height="100%"
    src="test-file-stub"
    width="100%"
  />
  <img
    className="nodeOverlay"
    height="100%"
    src="test-file-stub"
    width="100%"
  />
  <p>
    A_SERVICE
  </p>
</div>
`;

exports[`Node re-renders a node if service ui properties change 1`] = `
<div
  className="node"
  id="1"
  onDoubleClick={[Function]}
  style={
    Object {
      "left": 10,
      "top": 10,
    }
  }
>
  <img
    height="100%"
    src="test-file-stub"
    width="100%"
  />
  
  <p>
    A_SERVICE
  </p>
</div>
`;

exports[`Node re-renders a node if service ui properties change 2`] = `
<div
  className="node"
  id="1"
  onDoubleClick={[Function]}
  style={
    Object {
      "left": 600,
      "top": 500,
    }
  }
>
  <img
    height="100%"
    src="test-file-stub"
    width="100%"
  />
  
  <p>
    A_SERVICE
  </p>
</div>
`;

exports[`Node renders a node 1`] = `
<div
  className="node"
  id="1"
  onDoubleClick={[Function]}
  style={
    Object {
      "left": 10,
      "top": 10,
    }
  }
>
  <img
    height="100%"
    src="test-file-stub"
    width="100%"
  />
  
  <p>
    A_SERVICE
  </p>
</div>
`;
